File: backend/app/schemas/master_data.py
from decimal import Decimal
from typing import Literal
from uuid import UUID

from pydantic import BaseModel, ConfigDict, Field

from app.schemas.common import EntityRead

MaterialType = Literal["product", "semi_finished", "raw_material", "packing", "tooling"]
WorkCenterType = Literal["workstation", "equipment", "line", "inspection"]
WorkerType = Literal["operator", "inspector", "planner", "manager"]
MasterStatus = Literal["draft", "active", "inactive"]


class MaterialBase(BaseModel):
    code: str = Field(min_length=1, max_length=64)
    name: str = Field(min_length=1, max_length=128)
    spec: str | None = Field(default=None, max_length=255)
    unit: str = Field(min_length=1, max_length=32)
    material_type: MaterialType
    is_active: bool = True
    allow_empty_bom: bool = False
    remark: str | None = None


class MaterialCreate(MaterialBase):
    pass


class MaterialUpdate(BaseModel):
    model_config = ConfigDict(extra="forbid")

    name: str | None = Field(default=None, min_length=1, max_length=128)
    spec: str | None = Field(default=None, max_length=255)
    unit: str | None = Field(default=None, min_length=1, max_length=32)
    material_type: MaterialType | None = None
    is_active: bool | None = None
    allow_empty_bom: bool | None = None
    remark: str | None = None


class MaterialRead(MaterialBase, EntityRead):
    pass


class WorkCenterBase(BaseModel):
    code: str = Field(min_length=1, max_length=64)
    name: str = Field(min_length=1, max_length=128)
    work_center_type: WorkCenterType = "workstation"
    location: str | None = Field(default=None, max_length=128)
    is_active: bool = True
    remark: str | None = None


class WorkCenterCreate(WorkCenterBase):
    pass


class WorkCenterUpdate(BaseModel):
    model_config = ConfigDict(extra="forbid")

    name: str | None = Field(default=None, min_length=1, max_length=128)
    work_center_type: WorkCenterType | None = None
    location: str | None = Field(default=None, max_length=128)
    is_active: bool | None = None
    remark: str | None = None


class WorkCenterRead(WorkCenterBase, EntityRead):
    pass


class TeamBase(BaseModel):
    code: str = Field(min_length=1, max_length=64)
    name: str = Field(min_length=1, max_length=128)
    leader_name: str | None = Field(default=None, max_length=128)
    is_active: bool = True
    remark: str | None = None


class TeamCreate(TeamBase):
    pass


class TeamUpdate(BaseModel):
    model_config = ConfigDict(extra="forbid")

    name: str | None = Field(default=None, min_length=1, max_length=128)
    leader_name: str | None = Field(default=None, max_length=128)
    is_active: bool | None = None
    remark: str | None = None


class TeamRead(TeamBase, EntityRead):
    pass


class WorkerBase(BaseModel):
    code: str = Field(min_length=1, max_length=64)
    name: str = Field(min_length=1, max_length=128)
    worker_type: WorkerType = "operator"
    team_id: UUID | None = None
    is_active: bool = True
    remark: str | None = None


class WorkerCreate(WorkerBase):
    pass


class WorkerUpdate(BaseModel):
    model_config = ConfigDict(extra="forbid")

    name: str | None = Field(default=None, min_length=1, max_length=128)
    worker_type: WorkerType | None = None
    team_id: UUID | None = None
    is_active: bool | None = None
    remark: str | None = None


class WorkerRead(WorkerBase, EntityRead):
    pass


class OperationSkillOptionRead(BaseModel):
    operation_code: str
    operation_name: str


class WorkerOperationSkillRead(EntityRead):
    worker_id: UUID
    operation_code: str
    operation_name: str | None = None
    is_active: bool
    remark: str | None = None


class WorkerOperationSkillUpdate(BaseModel):
    model_config = ConfigDict(extra="forbid")

    operation_codes: list[str] = Field(default_factory=list)


class DefectReasonBase(BaseModel):
    code: str = Field(min_length=1, max_length=64)
    name: str = Field(min_length=1, max_length=128)
    category: str | None = Field(default=None, max_length=64)
    is_active: bool = True
    remark: str | None = None


class DefectReasonCreate(DefectReasonBase):
    pass


class DefectReasonUpdate(BaseModel):
    model_config = ConfigDict(extra="forbid")

    name: str | None = Field(default=None, min_length=1, max_length=128)
    category: str | None = Field(default=None, max_length=64)
    is_active: bool | None = None
    remark: str | None = None


class DefectReasonRead(DefectReasonBase, EntityRead):
    pass


class BomLineCreate(BaseModel):
    component_material_id: UUID
    line_no: int = Field(gt=0)
    qty_per: Decimal = Field(gt=0, max_digits=18, decimal_places=6)
    loss_rate: Decimal = Field(default=Decimal("0"), ge=0, max_digits=9, decimal_places=6)
    remark: str | None = None


class BomLineRead(BomLineCreate, EntityRead):
    component_material_code: str | None = None
    component_material_name: str | None = None

    model_config = ConfigDict(from_attributes=True)


class BomCreate(BaseModel):
    material_id: UUID
    version: str = Field(min_length=1, max_length=64)
    status: MasterStatus = "draft"
    remark: str | None = None
    lines: list[BomLineCreate] = Field(default_factory=list)


class BomUpdate(BaseModel):
    model_config = ConfigDict(extra="forbid")

    status: MasterStatus | None = None
    remark: str | None = None
    lines: list[BomLineCreate] | None = None


class BomRead(EntityRead):
    material_id: UUID
    version: str
    status: str
    remark: str | None = None
    lines: list[BomLineRead] = Field(default_factory=list)


class RoutingOperationCreate(BaseModel):
    seq: int = Field(gt=0)
    operation_code: str = Field(min_length=1, max_length=64)
    operation_name: str = Field(min_length=1, max_length=128)
    work_center_id: UUID
    setup_time_sec: int = Field(default=0, ge=0)
    unit_time_sec: int = Field(default=0, ge=0)
    is_active: bool = True
    remark: str | None = None


class RoutingOperationRead(RoutingOperationCreate, EntityRead):
    work_center_code: str | None = None
    work_center_name: str | None = None

    model_config = ConfigDict(from_attributes=True)


class RoutingCreate(BaseModel):
    material_id: UUID
    version: str = Field(min_length=1, max_length=64)
    status: MasterStatus = "draft"
    remark: str | None = None
    operations: list[RoutingOperationCreate] = Field(default_factory=list)


class RoutingUpdate(BaseModel):
    model_config = ConfigDict(extra="forbid")

    status: MasterStatus | None = None
    remark: str | None = None
    operations: list[RoutingOperationCreate] | None = None


class RoutingRead(EntityRead):
    material_id: UUID
    version: str
    status: str
    remark: str | None = None
    operations: list[RoutingOperationRead] = Field(default_factory=list)

File: frontend/src/types/masterData.ts
export type UUID = string

export type MaterialType = 'product' | 'semi_finished' | 'raw_material' | 'packing' | 'tooling'
export type WorkCenterType = 'workstation' | 'equipment' | 'line' | 'inspection'
export type WorkerType = 'operator' | 'inspector' | 'planner' | 'manager'
export type MasterStatus = 'draft' | 'active' | 'inactive'

export interface EntityRead {
  id: UUID
  tenant_id: UUID
  created_at: string
  updated_at: string
  deleted_at: string | null
}

export interface Page<T> {
  total: number
  items: T[]
}

export interface Material extends EntityRead {
  code: string
  name: string
  spec: string | null
  unit: string
  material_type: MaterialType
  is_active: boolean
  allow_empty_bom: boolean
  remark: string | null
}

export interface WorkCenter extends EntityRead {
  code: string
  name: string
  work_center_type: WorkCenterType
  location: string | null
  is_active: boolean
  remark: string | null
}

export interface Team extends EntityRead {
  code: string
  name: string
  leader_name: string | null
  is_active: boolean
  remark: string | null
}

export interface Worker extends EntityRead {
  code: string
  name: string
  worker_type: WorkerType
  team_id: UUID | null
  is_active: boolean
  remark: string | null
}

export interface OperationSkillOption {
  operation_code: string
  operation_name: string
}

export interface WorkerOperationSkill extends EntityRead {
  worker_id: UUID
  operation_code: string
  operation_name: string | null
  is_active: boolean
  remark: string | null
}

export interface DefectReason extends EntityRead {
  code: string
  name: string
  category: string | null
  is_active: boolean
  remark: string | null
}

export interface BomLine extends EntityRead {
  component_material_id: UUID
  component_material_code: string | null
  component_material_name: string | null
  line_no: number
  qty_per: string
  loss_rate: string
  remark: string | null
}

export interface Bom extends EntityRead {
  material_id: UUID
  version: string
  status: MasterStatus
  remark: string | null
  lines: BomLine[]
}

export interface RoutingOperation extends EntityRead {
  seq: number
  operation_code: string
  operation_name: string
  work_center_id: UUID
  work_center_code: string | null
  work_center_name: string | null
  setup_time_sec: number
  unit_time_sec: number
  is_active: boolean
  remark: string | null
}

export interface Routing extends EntityRead {
  material_id: UUID
  version: string
  status: MasterStatus
  remark: string | null
  operations: RoutingOperation[]
}

export type SimpleEntity = Material | WorkCenter | Team | Worker | DefectReason
